File: 2022/aoc2022/src/day11.rs
use std::fs;
use std::str::FromStr;

struct Monkey {
    inspected: u128,
    items: Vec<u128>,
    operation: Box<dyn Fn(u128) -> u128>,
    test_num: u128,
    if_true: usize,
    if_false: usize,
}
fn parse(s: &str) -> Monkey {
    let mut lines = s.lines();
    lines.next();
    let items_line = lines.next().unwrap();
    let items: Vec<_> = items_line
        .split_once("Starting items: ")
        .unwrap()
        .1
        .split(", ")
        .map(|x| u128::from_str(x).unwrap())
        .collect();
    let op: Box<dyn (Fn(u128) -> u128)> = match lines
        .next()
        .unwrap()
        .strip_prefix("  Operation: new = old ")
        .unwrap()
        .split_once(' ')
        .unwrap()
    {
        ("*", n) => {
            let num = u128::from_str(n).ok();
            if let Some(parsed_n) = num {
                Box::new(move |x: u128| x * parsed_n)
            } else {
                Box::new(|x: u128| x * x)
            }
        }

        ("+", n) => {
            let num = u128::from_str(n).ok();
            if let Some(parsed_n) = num {
                Box::new(move |x: u128| x + parsed_n)
            } else {
                unreachable!()
            }
        }
        _ => unreachable!(),
    };
    let test_num: u128 = lines
        .next()
        .unwrap()
        .strip_prefix("  Test: divisible by ")
        .map(|x| u128::from_str(x).unwrap())
        .unwrap();
    let if_true: usize = lines
        .next()
        .unwrap()
        .strip_prefix("    If true: throw to monkey ")
        .map(|x| usize::from_str(x).unwrap())
        .unwrap();
    let if_false: usize = lines
        .next()
        .unwrap()
        .strip_prefix("    If false: throw to monkey ")
        .map(|x| usize::from_str(x).unwrap())
        .unwrap();

    Monkey {
        inspected: 0,
        items,
        operation: op,
        test_num,
        if_true,
        if_false,
    }
}

pub fn run() {
    println!("day11");
    let s = fs::read_to_string("data/day11.txt").unwrap();

    let mut prod = 1;
    for (part, rounds) in [(1, 20), (2, 10000)] {
        let mut monkeys = Vec::new();
        for monkey_data in s.split("\n\n") {
            let monkey = parse(monkey_data);
            prod *= monkey.test_num;
            monkeys.push(monkey);
        }
        let num_monkeys = monkeys.len();
        for _ in 0..rounds {
            for ix in 0..num_monkeys {
                let mut inspected = 0;
                let mut moves = Vec::new();
                for item in &monkeys[ix].items {
                    inspected += 1;
                    let mut wl = (*monkeys[ix].operation)(*item);

                    wl = if part == 1 { wl / 3 } else { wl % prod };
                    let test_num = monkeys[ix].test_num;
                    if wl % test_num == 0 {
                        moves.push((monkeys[ix].if_true, wl))
                    } else {
                        moves.push((monkeys[ix].if_false, wl))
                    }
                }
                monkeys[ix].inspected += inspected;
                monkeys[ix].items = Vec::new();
                for (dest, wl) in moves {
                    monkeys[dest].items.push(wl)
                }
            }
        }
        let mut inspects = monkeys.iter().map(|m| (m.inspected)).collect::<Vec<_>>();
        inspects.sort_by(|a, b| b.cmp(a));
        println!(
            "PART {}: {} ({:?})",
            part,
            inspects[0] * inspects[1],
            inspects
        );
    }
}

File: 2022/aoc2022/src/day11.py
from dataclasses import dataclass
import re
from pathlib import Path
from typing import Callable


@dataclass
class Monkey:
    id: int
    items: [int]
    operation: Callable
    test: Callable
    dest: dict
    inspected: int

    def __repr__(self):
        return f"Monkey(id={self.id}, inspected={self.inspected})"


def parse(m: str) -> Monkey:
    global factors_prod
    lines = m.split("\n")
    id_ = int(re.findall(r'\d+', lines[0])[0])
    operation_parts = lines[2].split("= ")[-1].split(" ")

    def operation(old):
        op = operation_parts[1]
        left = old
        try:
            right = int(operation_parts[-1])
        except Exception:
            right = old
        if op == "*":
            return left * right
        if op == "+":
            return left + right
        raise ValueError("Invalid op %s, %s, %s", op, operation_parts, m)

    items = [int(k) for k in lines[1].split("items: ")[-1].split(", ")]

    def test(inp):
        return (inp % int(lines[3].split(" ")[-1]))
    factors_prod *= int(lines[3].split(" ")[-1])
    dest = {}
    dest[True] = int(lines[4].split(" ")[-1])
    dest[False] = int(lines[5].split(" ")[-1])
    return Monkey(id_, items, operation, test, dest, 0)


def setup():
    fname = Path(__file__).parent.absolute() / "../data/day11.txt"
    with open(fname) as f:
        data = f.read().strip().split("\n\n")
    monkeys = [parse(d) for d in data]
    return monkeys


def run(monkeys, op):
    for monkey in monkeys:
        items = monkey.items
        monkey.items = []
        monkey.inspected += len(items)
        for item in items:
            wl = monkey.operation(item)
            wl = op(wl)
            dest = monkey.dest[monkey.test(wl) == 0]
            monkeys[dest].items.append(wl)


def runs(rounds, monkeys, op):
    for i in range(rounds):
        run(monkeys, op)
    ms = sorted(monkeys, key=lambda x: x.inspected)[-2:]
    print(ms[0].inspected * ms[1].inspected)


if __name__ == "__main__":

    factors_prod = 1
    for n, op in zip([20, 10000], [lambda x: x // 3, lambda x: x % factors_prod]):
        monkeys = setup()
        runs(n, monkeys, op)
